File: objc/Project2/Project2/ViewController.m
//
//  ViewController.m
//  Project2
//
//  Created by Jinwoo Kim on 11/22/20.
//

#import "ViewController.h"

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.answer = @"";
    self.guesses = [@[] mutableCopy];
}


- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}

- (IBAction)submitGuess:(id)sender {
    NSString *guessString = self.guess.stringValue;
    NSMutableArray<NSString *> *guessArray = @[].mutableCopy;
    [guessString enumerateSubstringsInRange:NSMakeRange(0, guessString.length)
                                    options:NSStringEnumerationByComposedCharacterSequences
                                 usingBlock:^(NSString * _Nullable substring, NSRange substringRange, NSRange enclosingRange, BOOL * _Nonnull stop) {
                [guessArray addObject:substring];
    }];
    NSSet *guessSet = [NSSet setWithArray:guessArray];
    
//    NSLog(@"%@", guessSet);
    if ([guessSet count] != 4) return;
    if ([guessArray count] != 4) return;
    
    NSCharacterSet *badCharacters = [[NSCharacterSet characterSetWithCharactersInString:@"0123456789"] invertedSet];
    NSRange range = [guessString rangeOfCharacterFromSet:badCharacters];
    if (range.length != 0) return;
    
    [self.guesses insertObject:guessString atIndex:0];
    [self.tableView insertRowsAtIndexes:[NSIndexSet indexSetWithIndex:0] withAnimation:NSTableViewAnimationSlideDown];
    
    NSString *resultString = [self resultFor:guessString];
//    if ([guessArray containsObject:resultString]) {
        NSAlert *alert = [NSAlert new];
        alert.messageText = @"You win";
        alert.informativeText = @"Congratulations! Click OK to play again.";
        [alert runModal];
        [self startNewGame];
//    }
}

-(NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [self.guesses count];
}

- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
    NSTableCellView *vw = [tableView makeViewWithIdentifier:tableColumn.identifier owner:self];
    
    if (vw == nil) return nil;
    
    if ([tableColumn isEqual:@"Guess"]) {
        vw.textField.stringValue = self.guesses[row];
    } else {
        vw.textField.stringValue = [self resultFor:self.guesses[row]];
    }
    
    return vw;
}

- (BOOL)tableView:(NSTableView *)tableView shouldSelectRow:(NSInteger)row {
    return NO;
}

- (NSString *)resultFor:(NSString* )guess {
    __block int bulls = 0;
    __block int cows = 0;
    
    NSArray<NSString *> *guessLetters = [guess componentsSeparatedByString:@""];
    NSArray<NSString *> *answerLetters = [self.answer componentsSeparatedByString:@""];
    
    [guessLetters enumerateObjectsUsingBlock:^(NSString *letter, NSUInteger idx, BOOL *stop) {
        if ([letter isEqual:answerLetters[idx]]) {
            bulls += 1;
        } else if ([answerLetters containsObject:letter]) {
            cows += 1;
        }
    }];
    
    return [NSString stringWithFormat:@"%db %dc", bulls, cows];
}

- (void)startNewGame {
    self.guess.stringValue = @"";
    [self.guesses removeAllObjects];
    
    NSMutableArray<NSNumber *> *numbers = [@[] mutableCopy];
    
    for (int i=0; i<10; i++) {
        [numbers addObject:[NSNumber numberWithInt:i]];
    }
    
    for (int i=0; i<4; i++) {
        int randomIndex = arc4random_uniform((int)[numbers count]);
        int randomNumber = [numbers[randomIndex] intValue];
        [numbers removeObjectAtIndex:randomIndex];
        self.answer = [self.answer stringByAppendingFormat:@"%d", randomNumber];
    }
    
    [self.tableView reloadData];
}

@end

File: swift/Project2/Project2/ViewController.swift
//
//  ViewController.swift
//  Project2
//
//  Created by Jinwoo Kim on 11/19/20.
//

import Cocoa

class ViewController: NSViewController, NSTableViewDataSource, NSTableViewDelegate {
    @IBOutlet var tableView: NSTableView!
    @IBOutlet var guess: NSTextField!
    @IBOutlet var submitGuess: NSButton!
    
    var answer = ""
    var guesses = [String]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        startNewGame()
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }

    @IBAction func submitGuess(_ sender: Any) {
        let guessString = guess.stringValue
        
        guard Set(guessString).count == 4 else { return }
        guard guessString.count == 4 else { return }
        
        let badCharacters = CharacterSet(charactersIn: "0123456789").inverted
        guard guessString.rangeOfCharacter(from: badCharacters) == nil else { return }
        
        guesses.insert(guessString, at: 0)
        tableView.insertRows(at: IndexSet(integer: 0), withAnimation: .slideDown)
        
        let resultString = result(for: guessString)
//        if resultString.contains("4b") {
            let alert = NSAlert()
            alert.messageText = "You win"
            alert.informativeText = "Congratulations! Click OK to play again."
            alert.runModal()
            startNewGame()
//        }
    }
    
    func numberOfRows(in tableView: NSTableView) -> Int {
        return guesses.count
    }
    
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        guard let vw = tableView.makeView(withIdentifier: tableColumn!.identifier, owner: self) as? NSTableCellView else { return nil }
        if tableColumn?.title == "Guess" {
            // this is the "Guess" column; show a previous guess
            vw.textField?.stringValue = guesses[row]
        } else {
            // this is the "Result" columnl call our new method
            vw.textField?.stringValue = result(for: guesses[row])
        }
        
        return vw
    }
    
    func tableView(_ tableView: NSTableView, shouldSelectRow row: Int) -> Bool {
        return false
    }
    
    func result(for guess: String) -> String {
        var bulls = 0
        var cows = 0
        
        let guessLetters = Array(guess)
        let answerLetters = Array(answer)
        
        for (index, letter) in guessLetters.enumerated() {
            if letter == answerLetters[index] {
                bulls += 1
            } else if answerLetters.contains(letter) {
                cows += 1
            }
        }
        return "\(bulls)b \(cows)c"
    }
    
    func startNewGame() {
        guess.stringValue = ""
        guesses.removeAll()
        var numbers = Array(0...9)
        numbers.shuffle()
        for _ in 0..<4 {
            answer.append(String(numbers.removeLast()))
        }
        tableView.reloadData()
    }
}

